clang: integrate glean/lang/clang/tests/regression testsuite

alpmestan opened the pull request:
Current status: `Cases: 77  Tried: 77  Errors: 0  Failures: 3`

A breakdown of the failures is available below, not quite sure what the code is doing wrong for those, more pairs of eyes are very much welcome here :) @donsbot @simonmar

---

Failing tests
-------------

<details>
<summary>declarations/typeAlias1 (click to expand)</summary>

```
### Failure in: 0:2:declarations/typeAlias1
glean/test/regression/Glean/Regression/Snapshot.hs:202
declarationlocation.out: unexpected result
104c104
<               "columnEnd": 1
---
>               "columnEnd": 17
114c114
<         "columnEnd": 1
---
>         "columnEnd": 17

declarations.out: unexpected result
33c33
<               "columnEnd": 1
---
>               "columnEnd": 17
```

The problematic code is in `declarations/typeAlias1/test.mm`:

```
using U = id<Foo>;
```

compiled with `-xobjective-c++`. I tried adding `-std=c++{11, 14, 17}` here, out of despair, since those non-namespace `using` directives require at least C++11, to no avail. The `columnEnd` differences suggest that clang doesn't properly manage to parse this line and therefore that we're probably missing a magical flag to make this work.
</details>

<details>
<summary>declarations/objc-property1 (click to expand)</summary>

```
### Failure in: 0:8:declarations/objc-property1
glean/test/regression/Glean/Regression/Snapshot.hs:202
declarationlocation.out: unexpected result
61,93d60
<             "container": { "implementation": { "key": "A" } },
<             "signature": {
<               "key": { "returns": { "key": "int" }, "parameters": [ ] }
<             },
<             "isInstance": true,
<             "isOptional": false,
<             "isAccessor": true,
<             "source": {
<               "file": { "key": "test.m" },
<               "lineBegin": 19,
<               "columnBegin": 1,
<               "lineEnd": 19,
<               "columnEnd": 13
<             }
<           }
<         }
<       },
<       "source": {
<         "file": { "key": "test.m" },
<         "lineBegin": 19,
<         "columnBegin": 1,
<         "lineEnd": 19,
<         "columnEnd": 13
<       },
<       "name": "i"
<     }
<   },
<   {
<     "key": {
<       "decl": {
<         "objcMethod": {
<           "key": {
<             "selector": { "key": [ "i" ] },
119,156d85
<     }
<   },
<   {
<     "key": {
<       "decl": {
<         "objcMethod": {
<           "key": {
<             "selector": { "key": [ "setI" ] },
<             "container": { "implementation": { "key": "A" } },
<             "signature": {
<               "key": {
<                 "returns": { "key": "void" },
<                 "parameters": [
<                   { "name": { "key": "i" }, "type": { "key": "int" } }
<                 ]
<               }
<             },
<             "isInstance": true,
<             "isOptional": false,
<             "isAccessor": true,
<             "source": {
<               "file": { "key": "test.m" },
<               "lineBegin": 19,
<               "columnBegin": 1,
<               "lineEnd": 19,
<               "columnEnd": 13
<             }
<           }
<         }
<       },
<       "source": {
<         "file": { "key": "test.m" },
<         "lineBegin": 19,
<         "columnBegin": 1,
<         "lineEnd": 19,
<         "columnEnd": 13
<       },
<       "name": "setI"

declarations.out: unexpected result
2a3
>   { "key": [ ] },
114,211d114
<             }
<           }
<         }
<       },
<       {
<         "objcMethod": {
<           "key": {
<             "selector": { "key": [ "i" ] },
<             "container": { "implementation": { "key": "A" } },
<             "signature": {
<               "key": { "returns": { "key": "int" }, "parameters": [ ] }
<             },
<             "isInstance": true,
<             "isOptional": false,
<             "isAccessor": true,
<             "source": {
<               "file": { "key": "test.m" },
<               "lineBegin": 19,
<               "columnBegin": 1,
<               "lineEnd": 19,
<               "columnEnd": 13
<             }
<           }
<         }
<       },
<       {
<         "objcMethod": {
<           "key": {
<             "selector": { "key": [ "setI" ] },
<             "container": { "implementation": { "key": "A" } },
<             "signature": {
<               "key": {
<                 "returns": { "key": "void" },
<                 "parameters": [
<                   { "name": { "key": "i" }, "type": { "key": "int" } }
<                 ]
<               }
<             },
<             "isInstance": true,
<             "isOptional": false,
<             "isAccessor": true,
<             "source": {
<               "file": { "key": "test.m" },
<               "lineBegin": 19,
<               "columnBegin": 1,
<               "lineEnd": 19,
<               "columnEnd": 13
<             }
<           }
<         }
<       }
<     ]
<   },
<   {
<     "key": [
<       {
<         "objcMethod": {
<           "key": {
<             "selector": { "key": [ "i" ] },
<             "container": { "implementation": { "key": "A" } },
<             "signature": {
<               "key": { "returns": { "key": "int" }, "parameters": [ ] }
<             },
<             "isInstance": true,
<             "isOptional": false,
<             "isAccessor": true,
<             "source": {
<               "file": { "key": "test.m" },
<               "lineBegin": 19,
<               "columnBegin": 1,
<               "lineEnd": 19,
<               "columnEnd": 13
<             }
<           }
<         }
<       },
<       {
<         "objcMethod": {
<           "key": {
<             "selector": { "key": [ "setI" ] },
<             "container": { "implementation": { "key": "A" } },
<             "signature": {
<               "key": {
<                 "returns": { "key": "void" },
<                 "parameters": [
<                   { "name": { "key": "i" }, "type": { "key": "int" } }
<                 ]
<               }
<             },
<             "isInstance": true,
<             "isOptional": false,
<             "isAccessor": true,
<             "source": {
<               "file": { "key": "test.m" },
<               "lineBegin": 19,
<               "columnBegin": 1,
<               "lineEnd": 19,
<               "columnEnd": 13
```

The code:

```
@interface A {
  int _i;
}

@property (nonatomic, assign) int i;

@end

@implementation A

@synthesize i = _i;

@end
```

compiled with `-xobjective-c`. Not sure why we've got all the additional data with this branch compared to the expected output, it doesn't look like this is using anything special besides standard obj-c things (although I'm not Objective C expert).
</details>

<details>
<summary>xrefs/using-directive4 (click to expand)</summary>

```
### Failure in: 0:42:xrefs/using-directive4
glean/test/regression/Glean/Regression/Snapshot.hs:202
xrefs.out: unexpected result
1c1,238
< [ "@generated" ]
\ No newline at end of file
---
> [
>   "@generated",
>   {
>     "key": {
>       "file": { "key": "test.cpp" },
>       "xrefs": [
>         {
>           "range": { "begin": 396, "end": 403 },
>           "target": {
>             "declaration": {
>               "record_": {
>                 "key": {
>                   "name": {
>                     "key": {
>                       "name": { "key": "Promise" },
>                       "scope": { "global_": { } }
>                     }
>                   },
>                   "kind": { "struct_": { } },
>                   "source": {
>                     "file": { "key": "test.cpp" },
>                     "lineBegin": 13,
>                     "columnBegin": 1,
>                     "lineEnd": 25,
>                     "columnEnd": 1
>                   }
>                 }
>               }
>             }
>           }
>         },
>         {
>           "range": { "begin": 435, "end": 442 },
>           "target": {
>             "declaration": {
>               "record_": {
>                 "key": {
>                   "name": {
>                     "key": {
>                       "name": { "key": "Promise" },
>                       "scope": { "global_": { } }
>                     }
>                   },
>                   "kind": { "struct_": { } },
>                   "source": {
>                     "file": { "key": "test.cpp" },
>                     "lineBegin": 13,
>                     "columnBegin": 1,
>                     "lineEnd": 25,
>                     "columnEnd": 1
>                   }
>                 }
>               }
>             }
>           }
>         },
>         {
>           "range": { "begin": 472, "end": 476 },
>           "target": {
>             "declaration": {
>               "record_": {
>                 "key": {
>                   "name": {
>                     "key": {
>                       "name": { "key": "Task" },
>                       "scope": { "global_": { } }
>                     }
>                   },
>                   "kind": { "struct_": { } },
>                   "source": {
>                     "file": { "key": "test.cpp" },
>                     "lineBegin": 27,
>                     "columnBegin": 1,
>                     "lineEnd": 29,
>                     "columnEnd": 1
>                   }
>                 }
>               }
>             }
>           }
>         },
>         {
>           "range": { "begin": 618, "end": 625 },
>           "target": {
>             "declaration": {
>               "record_": {
>                 "key": {
>                   "name": {
>                     "key": {
>                       "name": { "key": "Promise" },
>                       "scope": { "global_": { } }
>                     }
>                   },
>                   "kind": { "struct_": { } },
>                   "source": {
>                     "file": { "key": "test.cpp" },
>                     "lineBegin": 13,
>                     "columnBegin": 1,
>                     "lineEnd": 25,
>                     "columnEnd": 1
>                   }
>                 }
>               }
>             }
>           }
>         },
>         {
>           "range": { "begin": 647, "end": 654 },
>           "target": {
>             "declaration": {
>               "record_": {
>                 "key": {
>                   "name": {
>                     "key": {
>                       "name": { "key": "Promise" },
>                       "scope": { "global_": { } }
>                     }
>                   },
>                   "kind": { "struct_": { } },
>                   "source": {
>                     "file": { "key": "test.cpp" },
>                     "lineBegin": 13,
>                     "columnBegin": 1,
>                     "lineEnd": 25,
>                     "columnEnd": 1
>                   }
>                 }
>               }
>             }
>           }
>         },
>         {
>           "range": { "begin": 682, "end": 686 },
>           "target": {
>             "declaration": {
>               "record_": {
>                 "key": {
>                   "name": {
>                     "key": {
>                       "name": { "key": "Task" },
>                       "scope": { "global_": { } }
>                     }
>                   },
>                   "kind": { "struct_": { } },
>                   "source": {
>                     "file": { "key": "test.cpp" },
>                     "lineBegin": 27,
>                     "columnBegin": 1,
>                     "lineEnd": 29,
>                     "columnEnd": 1
>                   }
>                 }
>               }
>             }
>           }
>         },
>         {
>           "range": { "begin": 722, "end": 724 },
>           "target": {
>             "declaration": {
>               "namespace_": {
>                 "key": {
>                   "name": { "key": { "name": { "key": "N0" } } },
>                   "source": {
>                     "file": { "key": "test.cpp" },
>                     "lineBegin": 31,
>                     "columnBegin": 1,
>                     "lineEnd": 35,
>                     "columnEnd": 1
>                   }
>                 }
>               }
>             }
>           }
>         },
>         {
>           "range": { "begin": 737, "end": 740 },
>           "target": {
>             "indirect": {
>               "key": {
>                 "via": {
>                   "usingDirective": {
>                     "key": {
>                       "name": {
>                         "key": {
>                           "name": { "key": "N0" },
>                           "scope": { "global_": { } }
>                         }
>                       },
>                       "source": {
>                         "file": { "key": "test.cpp" },
>                         "lineBegin": 40,
>                         "columnBegin": 3,
>                         "lineEnd": 40,
>                         "columnEnd": 20
>                       }
>                     }
>                   }
>                 },
>                 "target": {
>                   "declaration": {
>                     "function_": {
>                       "key": {
>                         "name": {
>                           "key": {
>                             "name": { "key": { "name": { "key": "foo" } } },
>                             "scope": {
>                               "namespace_": {
>                                 "key": { "name": { "key": "N0" } }
>                               }
>                             }
>                           }
>                         },
>                         "signature": {
>                           "key": {
>                             "returns": { "key": "Promise" },
>                             "parameters": [ ]
>                           }
>                         },
>                         "source": {
>                           "file": { "key": "test.cpp" },
>                           "lineBegin": 33,
>                           "columnBegin": 1,
>                           "lineEnd": 33,
>                           "columnEnd": 13
>                         }
>                       }
>                     }
>                   }
>                 }
>               }
>             }
>           }
>         }
>       ]
>     }
>   }
> ]
\ No newline at end of file
```

The problematic code is:

```
#include <coroutine>

struct Task;

struct Promise {
  bool await_ready() noexcept;
  template<typename T> void await_suspend(T) noexcept {}
  int await_resume() noexcept;

  Promise initial_suspend() noexcept;

  Promise final_suspend() noexcept;

  Task get_return_object() noexcept;
  void unhandled_exception() noexcept;
  void return_void() noexcept;
};

struct Task {
  using promise_type = Promise;
};

namespace N0 {

Promise foo();

}

namespace N1 {

Task bar() noexcept {
  using namespace N0;
  co_await foo();
}

}
```

compiled with `-std=c++14 -fcoroutines-ts`. Trying c++17 or c++20 seems to make
other tests fail without getting this one to succeed (and it _should_ succeed with `c++20` AFAICT, see e.g [this llvm ticket](https://bugs.llvm.org/show_bug.cgi?id=48172)).

The problem seems to be that we can't find `<coroutine>`, at least that's the
issue I'm facing when trying to compile the file manually.
And indeed, `find / -name 'coroutine'` only returns Boost's `coroutine` header
in the Docker container we're using. And this prevents the clang indexer from being able to do anything useful with this test file.
</details>

I'm going to move on to the other testsuites under `glean/lang/clang/tests` for now, while this patch gets reviewed. If anyone's got some idea about the failures, I'm all ears :-)

alpmestan commented on glean/lang/clang/tests/Glean/Regression/Driver/Clang.hs:
```diff
@@ -8,8 +8,13 @@
 
 main :: IO ()
-main = testMain Clang.driver
+main = getArgs >>= \args -> withArgs (args ++ ["--root", path]) $
```
OK, so I'll just make a separate module with the contents of this file, and only refer to that newer variant in the `.cabal` file. But the same problem will show up with all the deriving-oriented testsuites as well right? Do we want to essentially maintain two copies of each of those modules?

alpmestan replied:
Or perhaps we could have a single executable, to which we'd pass some string to specify which driver we want, and the corresponding "tests root" ?

simonmar commented on glean/lang/clang/tests/Glean/Regression/Driver/Clang.hs:
```diff
@@ -8,8 +8,13 @@
 
 main :: IO ()
-main = testMain Clang.driver
+main = getArgs >>= \args -> withArgs (args ++ ["--root", path]) $
```
We need some `--omit` now, right?

alpmestan commented:
Rebased and added a commit addressing your feedback. We now omit the tests that are failing in the CI env, in a separate executable as the one used by FB's internal infra.

alpmestan commented:
I rebased, but the tip of `main` seems to suffer from the same build error as this branch (as well as the other open PRs I have (#218, #220).
